### ISSUE TITLE
Not able to convert html table to PDF using PDFMake

### ISSUE THREAD

**Issue by kiranpandhure** [Aymkdn/html-to-pdfmake#34]
<table style="border-collapse:collapse; width:273pt; border:none" width="363"><colgroup></colgroup><colgroup><col style="width:19pt" width="25"></colgroup><colgroup><col style="width:159pt" width="212"></colgroup><colgroup><col style="width:95pt" width="126"></colgroup><tbody><tr height="19" style="height:14.25pt"><td class="xl77" height="82" rowspan="2" style="border:none; border-bottom:.5pt solid black; background:white; height:61.5pt; width:19pt; text-align:center; vertical-align:middle; border-top:1pt solid windowtext; border-right:1pt solid windowtext; border-left:1pt solid windowtext; padding:0px; white-space:nowrap" width="25"><span style="font-size:10pt"><span style="color:#305496"><span style="font-weight:700"><span style="font-family:Calibri,sans-serif"><span style="font-style:normal"><span style="text-decoration:none">Sr.</span></span></span></span></span></span></td><td class="xl79" rowspan="2" style="border:none; border-bottom:1.0pt solid black; background:white; width:159pt; text-align:center; vertical-align:middle; border-top:1pt solid windowtext; border-right:0.5pt solid windowtext; border-left:1pt solid windowtext; padding:0px; white-space:nowrap" width="212"><span style="font-size:10pt"><span style="color:#305496"><span style="font-weight:700"><span style="font-family:Calibri,sans-serif"><span style="font-style:normal"><span style="text-decoration:none">Title</span></span></span></span></span></span></td><td class="xl81" rowspan="2" style="border:none; border-bottom:1.0pt solid black; background:white; width:95pt; text-align:center; vertical-align:middle; border-top:1pt solid windowtext; border-right:0.5pt solid windowtext; border-left:0.5pt solid windowtext; white-space:normal; padding:0px" width="126"><span style="font-size:10pt"><span style="color:#305496"><span style="font-weight:700"><span style="font-family:Calibri,sans-serif"><span style="font-style:normal"><span style="text-decoration:none">DESC<br>?			( Description )</span></span></span></span></span></span></td></tr><tr height="63" style="height:47.25pt"></tr><tr height="42" style="height:31.5pt"><td class="xl70" height="84" rowspan="2" style="border:none; border-bottom:.5pt solid black; background:white; height:63.0pt; border-top:none; width:19pt; text-align:center; vertical-align:middle; border-right:0.5pt solid windowtext; border-left:0.5pt solid windowtext; white-space:normal; padding:0px" width="25"><span style="font-size:11pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">1</span></span></span></span></span></span></td><td class="xl72" rowspan="2" style="border:none; border-bottom:.5pt solid black; background:white; border-top:none; width:159pt; text-align:left; vertical-align:middle; border-right:0.5pt solid windowtext; border-left:0.5pt solid windowtext; white-space:normal; padding:0px" width="212"><span style="font-size:12pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">Test 1</span></span></span></span></span></span></td><td class="xl67" style="border:0.5pt solid windowtext; background:white; border-left:none; width:95pt; text-align:left; vertical-align:middle; white-space:normal; padding:0px" width="126"><span style="font-size:12pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">Test 1 Desc</span></span></span></span></span></span></td></tr><tr height="42" style="height:31.5pt"><td class="xl69" height="42" style="border:none; border-bottom:0.5pt solid windowtext; background:white; height:31.5pt; border-left:none; width:95pt; text-align:left; vertical-align:middle; border-top:none; border-right:0.5pt solid windowtext; white-space:normal; padding:0px" width="126"><span style="font-size:12pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">Test 1 Desc 2</span></span></span></span></span></span></td></tr><tr height="84" style="height:63.0pt"><td class="xl73" height="172" rowspan="2" style="border:none; border-bottom:.5pt solid black; background:white; height:129.0pt; width:19pt; text-align:center; vertical-align:middle; border-top:none; border-right:0.5pt solid windowtext; border-left:0.5pt solid windowtext; white-space:normal; padding:0px" width="25"><span style="font-size:11pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">2</span></span></span></span></span></span></td><td class="xl75" rowspan="2" style="border:none; border-bottom:.5pt solid black; background:white; border-top:none; width:159pt; text-align:center; vertical-align:middle; border-right:0.5pt solid windowtext; border-left:0.5pt solid windowtext; white-space:normal; padding:0px" width="212"><span style="font-size:12pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">Test 2</span></span></span></span></span></span></td><td class="xl67" style="border:0.5pt solid windowtext; background:white; border-top:none; border-left:none; width:95pt; text-align:left; vertical-align:middle; white-space:normal; padding:0px" width="126"><span style="font-size:12pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">Test 2 Desc</span></span></span></span></span></span></td></tr><tr height="88" style="height:66.0pt"><td class="xl69" height="88" style="border:none; border-bottom:0.5pt solid windowtext; background:white; height:66.0pt; border-left:none; width:95pt; text-align:left; vertical-align:middle; border-top:none; border-right:0.5pt solid windowtext; white-space:normal; padding:0px" width="126"><span style="font-size:12pt"><span style="color:#305496"><span style="font-family:Calibri,sans-serif"><span style="font-weight:400"><span style="font-style:normal"><span style="text-decoration:none">Test 2 Desc 2</span></span></span></span></span></span></td></tr></tbody></table>

**Comment by Aymkdn** [comment]
It would have been nice to use a greeting, followed by a sentence, and even maybe a "thank you" at some point.

This module is to transform HTML to PDFMake language, then PDFMake will do the hard work. The purpose of this tool is not to handle all CSS properties that exist today, especially when it becomes complicated design.

When I test your code, I can see you use a lot of CSS styles, and many of them are not compatible with what PDFMake support. Something as simple as `border:none` should be `border:0` to make it work with PDFMake.

Also you use `colgroup` and `col` that are not supported... and to make things even worse, you use plenty of unnacessary nested tags and styles.

Long story short: your code is way too much complicated to be magically handled by `html-to-pdfmake`. It could be greatly simplify to make it work.

**Comment by kiranpandhure** [comment]
Thank you for your response but even after removing **colgroup**  i am getting error

![image](https://user-images.githubusercontent.com/59874118/72350711-f5b68d80-3704-11ea-8cf8-955ad6bda988.png)

**Comment by Aymkdn** [comment]
As I said, `colgroup` is only one of the many problems.

> Long story short: your code is way too much complicated to be magically handled by `html-to-pdfmake`. It could be greatly simplify to make it work.

**Comment by kiranpandhure** [comment]
Thank for your response as per your findings i have removed colgroup and simplified html but still it gives me error of _calwidth as earlier i have posted with screenshot.
<table width="363">
    <tbody>
        <tr height="19">
            <td height="82" rowspan="2" width="25">Sr.</td>
            <td rowspan="2" width="212">Title</td>
            <td rowspan="2" width="126">DESC<br>?( Description )</td>
        </tr>
        <tr height="63"></tr>
        <tr height="42">
            <td height="84" rowspan="2" width="25">1</td>
            <td rowspan="2" width="212">Test 1</td>
            <td width="126">Test 1 Desc</td>
        </tr>
        <tr height="42">
            <td height="42" width="126">Test 1 Desc 2</td>
        </tr>
        <tr height="84">
            <td height="172" rowspan="2" width="25">2</td>
            <td rowspan="2" width="212">Test 2</td>
            <td width="126">Test 2 Desc</td>
        </tr>
        <tr height="88">
            <td height="88" width="126">Test 2 Desc 2</td>
        </tr>
    </tbody>
</table>

**Comment by Aymkdn** [comment]
You HTML code is still dirty.... The correct code is:
```html
<table width="363">
    <tbody>
        <tr height="19">
            <td height="82" width="25">Sr.</td>
            <td width="212">Title</td>
            <td width="126">DESC<br>?( Description )</td>
        </tr>
        <tr height="42">
            <td height="84" rowspan="2" width="25">1</td>
            <td rowspan="2" width="212">Test 1</td>
            <td width="126">Test 1 Desc</td>
        </tr>
        <tr height="42">
            <td height="42" width="126">Test 1 Desc 2</td>
        </tr>
        <tr height="84">
            <td height="172" rowspan="2" width="25">2</td>
            <td rowspan="2" width="212">Test 2</td>
            <td width="126">Test 2 Desc</td>
        </tr>
        <tr height="88">
            <td height="88" width="126">Test 2 Desc 2</td>
        </tr>
    </tbody>
</table>
```

That renders as the below:
<table width="363">
    <tbody>
        <tr height="19">
            <td height="82" width="25">Sr.</td>
            <td width="212">Title</td>
            <td width="126">DESC<br>?( Description )</td>
        </tr>
        <tr height="42">
            <td height="84" rowspan="2" width="25">1</td>
            <td rowspan="2" width="212">Test 1</td>
            <td width="126">Test 1 Desc</td>
        </tr>
        <tr height="42">
            <td height="42" width="126">Test 1 Desc 2</td>
        </tr>
        <tr height="84">
            <td height="172" rowspan="2" width="25">2</td>
            <td rowspan="2" width="212">Test 2</td>
            <td width="126">Test 2 Desc</td>
        </tr>
        <tr height="88">
            <td height="88" width="126">Test 2 Desc 2</td>
        </tr>
    </tbody>
</table>

And in the PDF:
![Capture](https://user-images.githubusercontent.com/946315/72413574-5f1fb600-3770-11ea-87bd-18bbb1febef6.JPG)

Please note that [PDFMake](https://pdfmake.github.io/docs/) doesn't support `width` and `height` on cells but it will only apply the same `width` or `height` to all columns/rows (see section "tables" from the [PDFMake Playground](http://pdfmake.org/playground.html)).

So, in your case it might be necessary to play with [special properties](https://github.com/Aymkdn/html-to-pdfmake#special-properties). Example:
```html
<!-- the important part is the `data-pdfmake` that defines the width for each column ([25,212,126]) and the height each row ([19,42,42,42,42]) ––>
<table data-pdfmake="{&quot;widths&quot;:[25,212,126],&quot;heights&quot;:[19,42,42,42,42]}">
    <tbody>
        <tr height="19">
            <td height="82" width="25">Sr.</td>
            <td width="212">Title</td>
            <td width="126">DESC<br>?( Description )</td>
        </tr>
        <tr height="42">
            <td height="84" rowspan="2" width="25">1</td>
            <td rowspan="2" width="212">Test 1</td>
            <td width="126">Test 1 Desc</td>
        </tr>
        <tr height="42">
            <td height="42" width="126">Test 1 Desc 2</td>
        </tr>
        <tr height="84">
            <td height="172" rowspan="2" width="25">2</td>
            <td rowspan="2" width="212">Test 2</td>
            <td width="126">Test 2 Desc</td>
        </tr>
        <tr height="88">
            <td height="88" width="126">Test 2 Desc 2</td>
        </tr>
    </tbody>
</table>
```

The result in the PDF:
![Capture](https://user-images.githubusercontent.com/946315/72414118-c4c07200-3771-11ea-9488-888e4ae79bf8.JPG)
